Circuit Breaker doesn't support dynamic fallback URI in Java configuration

Configure a circuit breaker in Java to copy the segments in the fallback URI

``
@Configuration
class RouteConfiguration {

	@Bean
    public RouteLocator gatewayRoutes(RouteLocatorBuilder builder) {
		
		builder.routes().route(r -> r.path("api/{*segments}").filters(f -> {
							    f.circuitBreaker(c -> c
											.setName("test")
											.setFallbackUri("forward:/inCaseOfFailureUseThis/{segments}")
											);
								}
								return f;
							})
							.uri(http://localhost:8080)
							);

``


Result:

Caused by: java.lang.IllegalArgumentException: Illegal character in path at index 32: forward:/inCaseOfFailureUseThis/{segments}
	at java.base/java.net.URI.create(URI.java:906) ~[na:na]


Caused by: java.net.URISyntaxException: Illegal character in path at index 32: forward:/inCaseOfFailureUseThis/{segments}
	at java.base/java.net.URI$Parser.fail(URI.java:2976) ~[na:na]
	at java.base/java.net.URI$Parser.checkChars(URI.java:3147) ~[na:na]
	at java.base/java.net.URI$Parser.parseHierarchical(URI.java:3229) ~[na:na]
	at java.base/java.net.URI$Parser.parse(URI.java:3177) ~[na:na]
	at java.base/java.net.URI.<init>(URI.java:623) ~[na:na]
	at java.base/java.net.URI.create(URI.java:904) ~[na:na]



Documentation extract

``
@Configuration
class RouteConfiguration {

    @Bean
    public RouterFunction<ServerResponse> gatewayRouterFunctionsCircuitBreakerFallback() {
        return route("circuitbreaker_route")
                .route(path("/consumingServiceEndpoint/{*segments}"), http("https://example.org"))
                .filter(circuitBreaker("myCircuitBreaker", URI.create("forward:/inCaseOfFailureUseThis/{segments}")))
                .build();
    }
}
``

Note:

Using the config via yaml is not possible in my case as a part of the URI (prefix) has to be done in the code.
So a configuration in Java of the fallback URI is required.

--- a/spring-cloud-gateway-server/src/main/java/org/springframework/cloud/gateway/filter/factory/SpringCloudCircuitBreakerFilterFactory.java
+++ b/spring-cloud-gateway-server/src/main/java/org/springframework/cloud/gateway/filter/factory/SpringCloudCircuitBreakerFilterFactory.java
@@ -17,6 +17,8 @@
 package org.springframework.cloud.gateway.filter.factory;
 
 import java.net.URI;
+import java.net.URLDecoder;
+import java.nio.charset.StandardCharsets;
 import java.util.HashSet;
 import java.util.List;
 import java.util.Set;
@@ -135,8 +137,10 @@ public abstract class SpringCloudCircuitBreakerFilterFactory
 
 			@Override
 			public String toString() {
+				String fallbackUriString = config.fallbackUri != null
+						? URLDecoder.decode(config.fallbackUri.toString(), StandardCharsets.UTF_8) : null;
 				return filterToStringCreator(SpringCloudCircuitBreakerFilterFactory.this)
-						.append("name", config.getName()).append("fallback", config.fallbackUri).toString();
+						.append("name", config.getName()).append("fallback", fallbackUriString).toString();
 			}
 		};
 	}
@@ -184,7 +188,7 @@ public abstract class SpringCloudCircuitBreakerFilterFactory
 		}
 
 		public Config setFallbackUri(String fallbackUri) {
-			return setFallbackUri(URI.create(fallbackUri));
+			return setFallbackUri(UriComponentsBuilder.fromUriString(fallbackUri).build().toUri());
 		}
 
 		public String getName() {

--- a/spring-cloud-gateway-server/src/test/java/org/springframework/cloud/gateway/filter/factory/SpringCloudCircuitBreakerFilterFactoryTests.java
+++ b/spring-cloud-gateway-server/src/test/java/org/springframework/cloud/gateway/filter/factory/SpringCloudCircuitBreakerFilterFactoryTests.java
@@ -92,6 +92,12 @@ public abstract class SpringCloudCircuitBreakerFilterFactoryTests extends BaseWe
 	}
 
 	@Test
+	public void testJavaConfigSegments() {
+		testClient.get().uri("/fallback/seg").exchange().expectStatus().isOk().expectBody()
+				.json("{\"from\":\"circuitbreakerfallbackcontrollerseg\"}");
+	}
+
+	@Test
 	public void filterErrorPage() {
 		testClient.get().uri("/delay/3").header("Host", "www.circuitbreakerconnectfail.org").accept(APPLICATION_JSON)
 				.exchange().expectStatus().is5xxServerError().expectBody().jsonPath("$.status").isEqualTo(500)

--- a/spring-cloud-gateway-server/src/test/java/org/springframework/cloud/gateway/filter/factory/SpringCloudCircuitBreakerTestConfig.java
+++ b/spring-cloud-gateway-server/src/test/java/org/springframework/cloud/gateway/filter/factory/SpringCloudCircuitBreakerTestConfig.java
@@ -83,6 +83,11 @@ public class SpringCloudCircuitBreakerTestConfig {
 		return Collections.singletonMap("from", "circuitbreakerfallbackcontroller3");
 	}
 
+	@GetMapping("/circuitbreakerFallbackController/seg")
+	public Map<String, String> fallbackcontrollerSeg() {
+		return Collections.singletonMap("from", "circuitbreakerfallbackcontrollerseg");
+	}
+
 	@GetMapping("/statusCodeFallbackController")
 	public Map<String, String> statusCodeFallbackController(ServerWebExchange exchange) {
 		return Collections.singletonMap("from", "statusCodeFallbackController");
@@ -106,6 +111,10 @@ public class SpringCloudCircuitBreakerTestConfig {
 		return builder.routes()
 				.route("fallback_controller_3",
 						r -> r.path("/fallback").filters(f -> f.setPath("/circuitbreakerFallbackController3")).uri(uri))
+				.route("fallback_with_segments", r -> r.path("/fallback/{*segments}")
+						.filters(f -> f.circuitBreaker(c -> c.setName("test")
+								.setFallbackUri("forward:/circuitbreakerFallbackController/{segments}")))
+						.uri(uri))
 				.route("circuitbreaker_fallback_forward", r -> r.host("**.circuitbreakerforward.org")
 						.filters(f -> f.circuitBreaker(config -> config.setFallbackUri("forward:/fallback"))).uri(uri))
 				.route("circuitbreaker_java",

